Source/Services/ReverseCallDispatcher.cs

```diff
@@ -3,195 +3,264 @@
-                };
-            _handleResponse = Task.WhenAny(_tasks);
         }
 
         /// <inheritdoc/>
-        public Task Call(TRequest request, Action<TResponse> callback)
-        {
-            return Call(request, response =>
-            {
-                callback(response);
-                return Task.CompletedTask;
-            });
-        }
+        public TConnectArguments Arguments { get; private set; }
 
         /// <inheritdoc/>
-        public Task Call(TRequest request, Func<TResponse, Task> callback)
+        public async Task<bool> ReceiveArguments(CancellationToken cancellationToken)
         {
-            lock (_lockObject)
+            if (await _clientStream.MoveNext(cancellationToken).ConfigureAwait(false))
             {
-                _lastCallNumber++;
-                _requestProperty.SetValue(request, _lastCallNumber);
-                _requests[_lastCallNumber] = callback;
-
-                var taskCompletionSource = new TaskCompletionSource<object>();
-                _callbackTasks[_lastCallNumber] = taskCompletionSource;
+                var arguments = _getConnectArguments(_clientStream.Current);
+                if (arguments != null)
+                {
+                    var callContext = _getArgumentsContext(arguments);
+                    if (callContext?.ExecutionContext != null)
+                    {
+                        _executionContextManager.CurrentFor(callContext.ExecutionContext.ToExecutionContext());
+                        Arguments = arguments;
+                        return true;
+                    }
+                    else
+                    {
+                        _logger.Warning("Received arguments, but call execution context was not set.");
+                    }
+                }
+                else
+                {
+                    _logger.Warning("Received initial message from client, but arguments was not set.");
+                }
+            }
 
-                _requestStream.WriteAsync(request).GetAwaiter().GetResult();
+            return false;
+        }
 
-                return taskCompletionSource.Task;
-            }
+        /// <inheritdoc/>
+        public async Task Accept(TConnectResponse response, CancellationToken cancellationToken)
+        {
+            var message = new TServerMessage();
+            _setConnectResponse(message, response);
+            await _serverStream.WriteAsync(message).ConfigureAwait(false);
+            await HandleClientMessages(cancellationToken).ConfigureAwait(false);
         }
 
         /// <inheritdoc/>
-        public async Task HandleCalls()
+        public Task Reject(TConnectResponse response, CancellationToken cancellationToken)
         {
-            await _handleResponse.ConfigureAwait(false);
-            var exception = _tasks.FirstOrDefault(_ => _.Exception != null)?.Exception;
-            if (exception != null) throw exception;
+            var message = new TServerMessage();
```

Comment:
So same as previous comment here...

Reply:
ad8f153 resolves this

Reply:
Not quite, you are still allowed to both `Accept` and `Reject` a connection. It should be either or. So just share the same bool and lock, instead of having two.

Reply:
Oops. Yes, that is true. Did not think about kalling both Reject and Accept

Reply:
This should resolve it https://github.com/dolittle-fundamentals/DotNET.Fundamentals/pull/310/commits/ea51a467e3a6db4176b8db55250bc631bdcf4bde

Reply:
Fixed by https://github.com/dolittle-fundamentals/DotNET.Fundamentals/pull/310/commits/43403fe1a3e9cbaf47d70b3e41f57d7ac5f3da9e